### PR TITLE
Clarify pointer arguments in libpe_rules

### PR DESCRIPTION
**Pull request by kgaillot**
... and a few libpe_status functions that snuck in

### CODE REVIEW

**Comment by kgaillot** [comment]
@nrwahl2 , can you review this when you get a chance? Low priority

File: lib/pengine/status.c [hunk 1]
```diff
@@ -439,18 +439,23 @@ pe_find_node_id(GList *nodes, const char *id)
...
 {
-    GList *gIter = nodes;
+    for (const GList *iter = nodes; iter != NULL; iter = iter->next) {
+        pe_node_t *node = (pe_node_t *) iter->data;
 
-    for (; gIter != NULL; gIter = gIter->next) {
-        pe_node_t *node = (pe_node_t *) gIter->data;
-
-        if (node && pcmk__str_eq(node->details->uname, uname, pcmk__str_casei)) {
+        if (pcmk__str_eq(node->details->uname, node_name, pcmk__str_casei)) {
```
Comment (nrwahl2) [inline lib/pengine/status.c:478]: Or we could drop the loop and call
```
g_list_find_custom(nodes, node_name, pcmk__strcase_equal)
```
and grab `->data` if the retval isn't NULL

Would have to make `pcmk__strcase_equal()` non-static though

File: lib/pengine/status.c [hunk 2]
```diff
@@ -423,19 +423,28 @@ pe_find_node_any(GList *nodes, const char *id, const char *uname)
...
-        pe_node_t *node = (pe_node_t *) gIter->data;
+    for (const GList *iter = nodes; iter != NULL; iter = iter->next) {
+        pe_node_t *node = (pe_node_t *) iter->data;
 
-        if (node && pcmk__str_eq(node->details->id, id, pcmk__str_casei)) {
+        /* @TODO Whether node IDs should be considered case-sensitive should
+         * probably depend on the node type, so functionizing the comparison
+         * would be worthwhile
+         */
+        if (pcmk__str_eq(node->details->id, id, pcmk__str_casei)) {
```
Comment (nrwahl2) [inline lib/pengine/status.c:457]: See previous comment re: `g_list_find_custom()` and `pcmk__strcase_equal()` for case-insensitive.

For case-sensitive, `g_strcmp0()` should work

File: lib/pengine/rules.c [hunk 3]
```diff
@@ -553,8 +553,17 @@ pe_unpack_nvpairs(xmlNode *top, xmlNode *xml_obj, const char *set_name,
                     always_first, overwrite, next_change);
 }
 
+/*!
+ * \brief Expand any regular expression submatches (%0-%9) in a string
```
Comment (nrwahl2) [inline lib/pengine/rules.c:546]: No issue with this per se, but the Pacemaker Explained doc could use an improved explanation (ideally with an example or two) of how regexes and submatches work with regard to rules. Examples would probably help elucidate the purpose of these utilities too.

File: lib/pengine/rules.c [hunk 4]
```diff
@@ -28,10 +28,10 @@ CRM_TRACE_INIT_DATA(pe_rules);
 /*!
  * \brief Evaluate any rules contained by given XML element
  *
- * \param[in]  xml          XML element to check for rules
- * \param[in]  node_hash    Node attributes to use when evaluating expressions
- * \param[in]  now          Time to use when evaluating expressions
- * \param[out] next_change  If not NULL, set to when evaluation will change
+ * \param[in,out] xml          XML element to check for rules
```
Comment (nrwahl2) [inline lib/pengine/rules.c:31]: *shakes fist at `getDocPtr()`*

File: lib/pengine/rules.c [hunk 5]
```diff
@@ -153,7 +153,7 @@ find_expression_type(xmlNode * expr)
  */
 
 static int
-phase_of_the_moon(crm_time_t * now)
+phase_of_the_moon(const crm_time_t *now)
```
Comment (nrwahl2) [inline lib/pengine/rules.c:156]: *shakes fist again for different reason*

File: lib/pengine/rules.c [hunk 6]
```diff
@@ -684,8 +694,18 @@ pe_eval_expr(xmlNode *rule, pe_rule_eval_data_t *rule_data, crm_time_t *next_cha
     return passed;
 }
 
+/*!
+ * \brief Evaluate rules
```
Comment (nrwahl2) [inline lib/pengine/rules.c:698]: It doesn't really matter (clear from name) but maybe "Evaluate a rule's subexpression" or something like that. We use "Evaluate rules" in the next commit for `pe_eval_rules()`.

Documentation is tricky due to the number of layers -- e.g., `pe_evaluate_rules()`, `pe_eval_rules()`, `pe_eval_expr()`, `pe_eval_subexpr()`, `pe_eval_<expr_type>()`...

File: lib/pengine/rules.c [hunk 7]
```diff
@@ -510,16 +510,6 @@ unpack_nvpair_blocks(xmlNode *top, const xmlNode *xml_obj, const char *set_name,
     }
 }
 
-void
-pe_eval_nvpairs(xmlNode *top, const xmlNode *xml_obj, const char *set_name,
-                pe_rule_eval_data_t *rule_data, GHashTable *hash,
-                const char *always_first, gboolean overwrite,
-                crm_time_t *next_change)
-{
-    unpack_nvpair_blocks(top, xml_obj, set_name, hash, always_first,
```
Comment (nrwahl2) [inline lib/pengine/rules.c:520]: Yeah this was odd

**Review by nrwahl2 — COMMENTED** [review]
Looks good

File: lib/pengine/status.c (hunk 1 above)
Comment (kgaillot) [reply lib/pengine/status.c:478]: Good idea, but not going to bother this time around

File: lib/pengine/rules.c (hunk 3 above)
Comment (kgaillot) [reply lib/pengine/rules.c:546]: Definitely overdue, added commit on top. Generated docs viewable at https://people.redhat.com/~kgaillot/pacemaker/doc/

File: lib/pengine/rules.c (hunk 6 above)
Comment (kgaillot) [reply lib/pengine/rules.c]: Oh yeah, I meant to come back to that. Added commit on top.

File: doc/sphinx/Pacemaker_Explained/rules.rst [hunk 8]
```diff
@@ -765,6 +765,60 @@ would have its preference to run the resource increased by ``1234`` whereas
...
+
+Location constraints may use ``rsc-pattern`` to apply the constraint to all
+resources whose IDs match the given pattern (see :ref:`s-rsc-pattern`). The
+pattern may contain up to 9 submatches in parentheses, whose values may be used
+as ``%1`` through ``%9`` in a rule's ``score-attribute`` or a rule expression's
+``attribute``.
+
+As an example, the following configuration (only relevant parts are shown)
+gives the resources **server-httpd** and **ip-httpd** a preference of 100 on
+**node1** and 50 on **node2**, and **ip-gateway** a prefernce of -100 on
```
Comment (nrwahl2) [inline doc/sphinx/Pacemaker_Explained/rules.rst:781]: s/prefernce/preference/

File: doc/sphinx/Pacemaker_Explained/rules.rst [hunk 9]
```diff
@@ -765,6 +765,60 @@ would have its preference to run the resource increased by ``1234`` whereas
...
+         </node>
+         <node id="2" uname="node2">
+            <instance_attributes id="node2-attrs">
+               <nvpair id="node2-prefer-httpd" name="prefer-httpd" value="50"/>
+               <nvpair id="node2-prefer-gateway" name="prefer-gateway" value="200"/>
+            </instance_attributes>
+         </node>
+      </nodes>
+      <resources>
+         <primitive id="server-httpd" class="ocf" provider="heartbeat" type="apache" />
```
Comment (nrwahl2) [inline doc/sphinx/Pacemaker_Explained/rules.rst:803]: Completely unimportant but while we're here, some of these have a space before the closing `/>` and others don't. Might as well line it up when fixing the typo.

**Review by nrwahl2 — COMMENTED** [review]
Thanks! Looks great. Once we fix the typo, good to go.

File: doc/sphinx/Pacemaker_Explained/rules.rst (hunk 9 above)
Comment (kgaillot) [reply doc/sphinx/Pacemaker_Explained/rules.rst]: Thanks, updated commit for both of these

**Pull request merged**